mypy/subtypes.py

```diff
@@ -303,6 +331,227 @@ def visit_type_type(self, left: TypeType) -> bool:
         return False
 
 
+@contextmanager
+def pop_on_exit(stack: List[Tuple[Instance, Instance]],
+                left: Instance, right: Instance) -> Iterator[None]:
+    stack.append((left, right))
+    yield
+    stack.pop()
+
+
+def is_protocol_implementation(left: Instance, right: Instance, allow_any: bool = True) -> bool:
+    """Check whether 'left' implements the protocol 'right'. If 'allow_any' is False, then
+    check for a proper subtype. Treat recursive protocols by using the 'assuming'
+    structural subtype matrix (in sparse representation), see comment in nodes.TypeInfo.
+    """
+    assert right.type.is_protocol
+    assuming = right.type.assuming if allow_any else right.type.assuming_proper
+    for (l, r) in reversed(assuming):
+        if sametypes.is_same_type(l, left) and sametypes.is_same_type(r, right):
+            return True
+    with pop_on_exit(assuming, left, right):
+        if right.type.protocol_members is None:
+            # This type has not been yet analyzed, probably a call from make_simplified_union.
+            return False
+        for member in right.type.protocol_members:
+            # nominal subtyping currently ignores '__init__' and '__new__' signatures
+            if member in ('__init__', '__new__'):
+                continue
+            supertype = find_member(member, right, left)
```

Comment:
Add a comment here explaining that the third argument is to bind self, and that we always bind self-type to subtype.